module/zfs/dmu.c

```diff
@@ -547,19 +556,19 @@ dmu_buf_hold_array_by_dnode(dnode_t *dn, uint64_t offset, uint64_t length,
 		}
 
 		/* initiate async i/o */
-		if (read)
+		if (read) {
 			(void) dbuf_read(db, zio, dbuf_flags);
+			if (db->db_state != DB_CACHED)
+				missed = B_TRUE;
```

Comment:
I'm guessing that dbuf_read() usually doesn't initiate any io, or change db_state, until we wait for the parent `zio`.  Even so, it seems more intuitive to check the cached state before "doing" the read (calling dbuf_read()), because conceptually once the dbuf is read, its state is always cached.

Reply:
There are cases when dbuf_read() changes the dbuf state to DB_CACHED, such as hole reads.  And I think cases of DMU miss, but ARC hit should get there too, etc.  And if we got the data without waiting, I don't really care where they are coming from, prefetch would be just a waste of time.

Reply:
I see.  In that case could you add a comment explaining the ordering here?  Something like `We check the db_state after calling dbuf_read because (1) dbuf_read() may change the state to CACHED due to a hit in the ARC, and (2) on a cache miss, a child will have been added to "zio" but not yet issued, so the state will not yet be CACHED.`

Reply:
Added.